source/adapters/level_zero/command_buffer.cpp

```diff
@@ -41,13 +41,29 @@
   This WaitEvent is reset at the end of the suffix, along with reset commands
   to reset the L0 events used to implement the UR sync-points.
 
+  The prefix has been split into two command-lists for performance concerns.
+  1) This allows the `waitEvent` to be signaled directly on the host if 
+  the waiting list is empty, thus avoiding the need to submit a cmd list.
+  2) Enqueuing a reset ZE command for all events in the command-buffer is time
+  consumming, especially for large graphs.
+  However, this task is not needed for every submission, but only once, when the
+  command-buffer is fixed, i.e. when the command-buffer is finalized. The
+  decorelation between the reset cmd-list and the wait command-list allow us to
+  create and enqueue the reset commands when finalizing the command-buffer, 
+  and only create the wait command-list at submission.
+
   ┌──────────┬────────────────────────────────────────────────┬─────────┐
   │  Prefix  │ Commands added to UR command-buffer by UR user │ Suffix  │
   └──────────┴────────────────────────────────────────────────┴─────────┘
```

Comment:
A deficiency of the existing documentation is that it doesn't show if prefix and suffix are commands added to the same command-list as the main command-list, or new command-lists.

e.g  `SignalCommandList` is a new command-list, but the documentation doesn't really show that.

Would you be able to do a more general improvement to this doc that makes it clear when "prefix" and "suffix" mean new command-lists or commands to an existing command-list? Feel free to define new terminology for this instead.

As an aside, I also wonder if we could move the design doc sections/diagrams related to UR implementation into this repo from DPC++ and link it, so that we could avoid having this long comment and just link to it from the DPC++ doc.

Reply:
`command_buffer.cpp` now points to the DPC++ doc, which has been improved in this PR: https://github.com/reble/llvm/pull/357

Reply:
sounds good.